File: LegacySource/settlement/room/law/court/CourtStation.cs
using System;
using snake2d.util.bit;
using snake2d.util.datatypes;
using snake2d.util.datatypes;

namespace settlement.room.law.court
{
    public sealed class CourtStation
    {
        private static readonly Bit workReserved = new Bit(0b010000);
        private static readonly Bits state = new Bits(0b01111);
        private const int STATE_RESERVABLE = 0;
        private const int STATE_RESERVED = 1;
        private const int STATE_USED = 2;
        private const int STATE_JUDGING = 3;

        private readonly Coo cooCriminal = new Coo();
        private readonly Coo cooJudge = new Coo();
        private int rot;
        private int data;
        private CourtInstance ins;
        private static readonly CourtStation self = new CourtStation();

        public static bool IsJudge(COORDINATE c)
        {
            int s = SETT.ROOMS().fData.tileData.Get(c.x(), c.y());
            return s == Constructor.CodeWork;
        }

        public static CourtStation Init(int tx, int ty)
        {
            CourtInstance ins = SETT.ROOMS().COURT.Get(tx, ty);
            if (ins == null)
                return null;
            int c = SETT.ROOMS().fData.tileData.Get(tx, ty);
            self.ins = ins;
            if (c == Constructor.CodeCriminal)
            {
                self.cooCriminal.Set(tx, ty);
                self.rot = SETT.ROOMS().fData.item.Get(tx, ty).rotation;
                DIR d = DIR.ORTHO.Get(self.rot);
                self.cooJudge.Set(tx + d.x() * Constructor.Distance, ty + d.y() * Constructor.Distance);
                self.data = SETT.ROOMS().data.Get(self.cooJudge);
                return self;
            }
            if (c == Constructor.CodeWork)
            {
                self.cooJudge.Set(tx, ty);
                self.rot = SETT.ROOMS().fData.item.Get(tx, ty).rotation;
                DIR d = DIR.ORTHO.Get(self.rot).Perpendicular();
                self.cooCriminal.Set(tx + d.x() * Constructor.Distance, ty + d.y() * Constructor.Distance);
                self.data = SETT.ROOMS().data.Get(self.cooJudge);
                return self;
            }
            return null;
        }

        public DIR CriminalDir()
        {
            return DIR.ORTHO.Get(rot);
        }

        public DIR JudgeDir()
        {
            return DIR.ORTHO.Get(rot).Perpendicular();
        }

        public bool CriminalReservedCanBe()
        {
            return state.Get(data) == STATE_RESERVABLE;
        }

        private void CriminalReserve()
        {
            if (!CriminalReservedCanBe())
                throw new RuntimeException();
            data = state.Set(data, STATE_RESERVED);
            Save();
        }

        public bool CriminalReserved()
        {
            return state.Get(data) >= STATE_RESERVED;
        }

        public bool CriminalIsUsing()
        {
            return state.Get(data) == STATE_USED;
        }

        public void CriminalUse()
        {
            data = state.Set(data, STATE_USED);
            Save();
        }

        public void CriminalClear()
        {
            data = state.Set(data, STATE_RESERVABLE);
            Save();
        }

        private bool WorkReservedCanBe()
        {
            return !workReserved.Is(data) && state.Get(data) > STATE_RESERVED;
        }

        private void WorkReserve()
        {
            data = workReserved.Set(data);
            Save();
        }

        public void WorkUse()
        {
            data = state.Set(data, STATE_JUDGING);
            Save();
        }

        public bool CriminalIsBeingHeard()
        {
            return state.Get(data) == STATE_JUDGING;
        }

        public void WorkCancel()
        {
            data = workReserved.Clear(data);
            Save();
        }

        public bool WorkReserved()
        {
            if (state.Get(data) <= STATE_RESERVED)
                WorkCancel();
            return workReserved.Is(data);
        }

        private void Save()
        {
            Add(SETT.ROOMS().data.Get(cooJudge.x(), cooJudge.y()), -1);
            Add(data, 1);
            SETT.ROOMS().data.Set(ins, cooJudge, data);
        }

        private void Add(int data, int delta)
        {
            if (state.Get(data) != STATE_RESERVABLE)
                ins.Inc(delta, 0);
            if (!workReserved.Is(data) && state.Get(data) > STATE_RESERVED)
                ins.Inc(0, delta);
        }

        public COORDINATE CooJudge()
        {
            return cooJudge;
        }

        public COORDINATE CooCriminal()
        {
            return cooCriminal;
        }
    }
}

File: settlement/room/law/court/CourtStation.java
package settlement.room.law.court;

import settlement.main.SETT;
import snake2d.util.bit.Bit;
import snake2d.util.bit.Bits;
import snake2d.util.datatypes.COORDINATE;
import snake2d.util.datatypes.Coo;
import snake2d.util.datatypes.DIR;

public final class CourtStation{

	private final static Bit workReserved = new Bit(0b010000);
	private final static Bits state = new Bits(0b01111);
	private final static int STATE_RESERVABLE = 0;
	private final static int STATE_RESERVED = 1;
	private final static int STATE_USED = 2;
	private final static int STATE_JUDGING = 3;
	
	private final Coo cooCriminal = new Coo();
	private final Coo cooJudge = new Coo();
	private int rot;
	private int data;
	private CourtInstance ins;
	private static final CourtStation self = new CourtStation();

	
	static boolean isJudge(COORDINATE c) {
		int s = SETT.ROOMS().fData.tileData.get(c.x(), c.y());
		return s == Constructor.codeWork;
	}
	
	static CourtStation init(int tx, int ty) {
		CourtInstance ins = SETT.ROOMS().COURT.get(tx, ty);
		if (ins == null)
			return null;
		int c = SETT.ROOMS().fData.tileData.get(tx, ty);
		self.ins = ins;
		if (c == Constructor.codeCriminal) {
			self.cooCriminal.set(tx, ty);
			self.rot = SETT.ROOMS().fData.item.get(tx, ty).rotation;
			DIR d = DIR.ORTHO.get(self.rot);
			self.cooJudge.set(tx+d.x()*Constructor.distance, ty+d.y()*Constructor.distance);
			self.data = SETT.ROOMS().data.get(self.cooJudge);
			return self;
		}
		if (c == Constructor.codeWork) {
			self.cooJudge.set(tx, ty);
			self.rot = SETT.ROOMS().fData.item.get(tx, ty).rotation;
			DIR d = DIR.ORTHO.get(self.rot).perpendicular();
			self.cooCriminal.set(tx+d.x()*Constructor.distance, ty+d.y()*Constructor.distance);
			self.data = SETT.ROOMS().data.get(self.cooJudge);
			return self;
		}
		return null;
	}
	
	public DIR criminalDir() {
		return DIR.ORTHO.get(rot);
	}
	
	public DIR jundgeDir() {
		return DIR.ORTHO.get(rot).perpendicular();
	}
	
	public boolean criminalReseveredCanBe() {
		return state.get(data) == STATE_RESERVABLE;
	}
	
	void criminalReserve() {
		if (!criminalReseveredCanBe())
			throw new RuntimeException();
		data = state.set(data, STATE_RESERVED);
		save();
	}
	
	public boolean criminalReserved() {
		return state.get(data) >= STATE_RESERVED;
	}
	
	public boolean criminalIsUsing() {
		return state.get(data) == STATE_USED;
	}
	
	public void criminalUse() {
		data = state.set(data, STATE_USED);
		save();
	}

	public void criminalClear() {
		data = state.set(data, STATE_RESERVABLE);
		save();
	}
	
	boolean workReservedCanBe() {
		return !workReserved.is(data) && state.get(data) > STATE_RESERVED;
	}
	
	void workReserve() {
		data = workReserved.set(data);
		save();
	}
	
	public void workUse() {
		data = state.set(data, STATE_JUDGING);
		save();
	}
	
	public boolean criminalIsBeeingHeard() {
		return state.get(data) == STATE_JUDGING;
	}

	public void workCancel() {
		data = workReserved.clear(data);
		save();
	}
	
	public boolean workReserved() {
		if (state.get(data) <= STATE_RESERVED)
			workCancel();
		return workReserved.is(data);
	}	
	
	private void save() {
		add(SETT.ROOMS().data.get(cooJudge.x(), cooJudge.y()), -1);
		add(data, 1);
		SETT.ROOMS().data.set(ins, cooJudge, data);
	}
	
	private void add(int data, int delta) {
		if (state.get(data) != STATE_RESERVABLE)
			ins.inc(delta, 0);
		if (!workReserved.is(data) && state.get(data) > STATE_RESERVED)
			ins.inc(0, delta);
	}
	
	public COORDINATE cooJudge() {
		return cooJudge;
	}

	public COORDINATE cooCriminal() {
		return cooCriminal;
	}

	
}
